src/crystal/system/win32/event_loop_iocp.cr

```diff
@@ -1,14 +1,30 @@
+require "c/iocp"
 require "crystal/system/print_error"
 
 module Crystal::EventLoop
-  @@queue = Deque(Fiber).new
+  @@queue = Deque(Crystal::Event).new
 
   # Runs the event loop.
   def self.run_once : Nil
-    next_fiber = @@queue.pop?
+    unless @@queue.empty?
+      next_event = @@queue.min_by { |e| e.wake_in }
+      time_elapsed = (Time.monotonic - next_event.slept_at)
 
-    if next_fiber
-      Crystal::Scheduler.enqueue next_fiber
+      unless time_elapsed > next_event.wake_in
+        sleepy_time = (next_event.wake_in - time_elapsed).total_milliseconds.to_i
+        io_entry = Slice.new(1, LibC::OVERLAPPED_ENTRY.new)
+
+        if LibC.GetQueuedCompletionStatusEx(Thread.current.iocp, io_entry, 1, out removed, sleepy_time, false)
+          if removed == 1 && io_entry.first.lpOverlapped
+            next_event = io_entry.first.lpOverlapped.value.cEvent.unsafe_as(Crystal::Event)
+          end
```

Comment:
If removed is 0, this means that the sleepy_time passed without hitting any completed events, right? If so, if I read the code further down the min_value above is enqueued in the scheduler at line 27. But does anything actually cancel the completion of the event or keep track that it has timed out? Will it continue and eventually be completed? Because if it will then you will have a strange situation potentially enqueuing a fiber in unknown status unless I'm totally confused.

Reply:
>If removed is 0, this means that the sleepy_time passed without hitting any completed events, right?

Not exactly. If we make the GetQueuedCompletionStatusEx call, it times out, returns true, but removed is 0, then that means a sleep event completed. The idea was to treat sleep events the same as real i/o and use GetQueuedCompletionStatusEx's timeout as the mechanism for completing sleep events.

>If so, if I read the code further down the min_value above is enqueued in the scheduler at line 27. But does anything actually cancel the completion of the event or keep track that it has timed out? 

Yes, we then signal completion by removing the event from the event loop queue and telling the scheduler to switch to the event's associated fiber.

> Will it continue and eventually be completed? Because if it will then  you will have a strange situation potentially enqueuing a fiber in  unknown status unless I'm totally confused.

As another example, let's say we're in the middle of a 5 second sleep event blocking on the GetQueuedCompletionStatusEx call and some real i/o (like a socket or file read) comes in. In this case,  GetQueuedCompletionStatusEx will return as soon as possible (likely before our 5 second sleep is up) with the information about the real i/o. The `removed` arg will be 1 and we'll have a pointer to `lpOverlapped` which contains the crystal event/fiber that started the real i/o. Since the 5 second sleep was interrupted, we can safely put aside handling the sleep event and instead set the real i/o as the `next_event`, dequeue it from the event loop, and tell the scheduler to run that associated fiber. 

The sleep event that was put aside will get handled on the next iteration after we calculate `time_elapsed` (which will likely be a hair over 5 seconds), at which point we completely skip the GetQueuedCompletionStatusEx call completely and instead enqueue the fiber associated with the sleep event. Hopefully that example adds more context.

Reply:
> use GetQueuedCompletionStatusEx's timeout as the mechanism for completing sleep events

I see. I guess I got confused by having spent quite a lot of time staring at a different completion mechanism on linux (io_uring) that do support arbitrary wait events directly.  

Ok, I don't know enough about windows to say if that or using `CreateTimerQueueTimer` or something similar would make most sense. But it makes a lot more sense than my initial impression at least. :)

How would event timeouts work? I suppose there is some way of putting that information in the entry on submission, but how would the result be communicated and would there need to be some handling in here for that?

Reply:
>How would event timeouts work? I suppose there is some way of putting that information in the entry on submission, but how would the result be communicated and would there need to be some handling in here for that?

I missed this. After `GetQueuedCompletionStatusEx` is called and times out, it could be a sleep event completing or a I/O timeout, but I'm not handling that second case. There's an IO::Evented instance passed into the event with a read/write_timed_out property on it. 

I will add some logic in the event loop to check if `next_event` was a read/write event and if so set that boolean appropriately and thereby communicate the result.